tests/test-performance-async-publish-actions.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Automattic\VIP\Performance;
+
+class async_publish_actions_Test extends \WP_UnitTestCase {
+	/**
+	 * Prepare test environment
+	 */
+	function setUp() {
+		parent::setUp();
+
+		// make sure the schedule is clear.
+		_set_cron_array( array() );
+	}
+
+	/**
+	 * Clean up after our tests
+	 */
+	function tearDown() {
+		// make sure the schedule is clear.
+		_set_cron_array( array() );
+
+		parent::tearDown();
+	}
+
+	/**
+	 * Test that offload is scheduled when post is published
+	 */
+	function test_schedule_offload() {
+		$post = [
+			'post_title'   => 'Tommy Tutone - 867-5309/Jenny',
+			'post_content' => 'https://www.youtube.com/watch?v=6WTdTwcmxyo',
+			'post_status'  => 'draft',
+		];
+
+		$pid = wp_insert_post( $post, true );
+
+		// Insert succeeded.
+		$this->assertFalse( is_wp_error( $pid ) );
+
+		$args = [
+			'post_id'    => (int) $pid,
+			'new_status' => 'draft',
+			'old_status' => 'new',
+		];
+
+		$next_transition    = wp_next_scheduled( ASYNC_TRANSITION_EVENT, $args );
+
+		// No async event for a draft post.
+		$this->assertFalse( $next_transition );
+
+		wp_publish_post( $pid );
+
+		$args = [
+			'post_id'    => (int) $pid,
+			'new_status' => 'publish',
+			'old_status' => 'draft',
+		];
+
+		$next_transition    = wp_next_scheduled( ASYNC_TRANSITION_EVENT, $args );
+
+		// Event scheduled after publish.
+		$this->assertTrue( is_int( $next_transition ) );
```

Comment:
Ideally, this would be split off into multiple tests. A good rule of thumb is to avoid multiple state changes + multiple unrelated assert calls in a single test. Each code path + varying input should have its own tests.

Another good rule of thumb is that if you do end up with multiple assert calls in a single test, you should include the message parameter to add clues as which assertion fails (instead of just getting an error that says `expected true was false`.

Anyway, we could break these off into the following tests:

- Test auto-draft post doesn't schedule.
- Test already scheduled post doesn't schedule.
- Test when both new and old status are the same.
- Test when new status is publish.
- Test when old status is publish.
- Test that publishing a post and then another post schedules different events.

Reply:
795e7b and ee0c038